File: src/advent2020/day2/part2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func isValidPassword(line string) bool {
	if len(line) <= 0 {
		return false
	}

	spl := strings.Split(line, " ")
	first, err := strconv.Atoi(spl[0])
	last, err := strconv.Atoi(spl[1])
	check(err)

	char := spl[2]
	pass := spl[3]

	hasFirst := string(pass[first-1]) == char
	hasLast := string(pass[last-1]) == char

	return (hasFirst && !hasLast) || (!hasFirst && hasLast)
}

func main() {
	data, err := ioutil.ReadFile("./input.txt")
	check(err)

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	count := 0
	for i := 0; i < len(lines); i++ {
		line := strings.ReplaceAll(strings.ReplaceAll(lines[i], ": ", " "), "-", " ")

		if isValidPassword(line) {
			count++
		}
	}
	fmt.Printf("%v\n", count)
}

File: src/advent2020/day2/part2.js
//const input = (await Deno.readTextFile("./input_light.txt")).split("\n");
const input = (await Deno.readTextFile("./input.txt")).split("\n");

function isValidPassword(line) {
    if (!line) return false;
    const [rule, pass] = line.split(': ');
    const [mm, char] = rule.split(' ');
    const [min, max] = mm.split('-').map(n => parseInt(n, 10));

    console.log({min, max, char, pass});
    const countChars = pass.split('').filter(c => c === char).length;


    return countChars <= max && countChars >= min;
}

function isValidPassword2(line) {
    if (!line) return false;
    const [rule, pass] = line.split(': ');
    const [mm, char] = rule.split(' ');
    const [first, last] = mm.split('-').map(n => parseInt(n, 10));

    const hasFirst = pass.split('')[first -1] === char;
    const hasLast = pass.split('')[last - 1] === char;

    console.log({first, last, char, pass});

    return (hasFirst && !hasLast) || (!hasFirst && hasLast);
}

console.log(input.map(isValidPassword).filter(Boolean).length);
console.log(input.map(isValidPassword2).filter(Boolean).length);
